gh-107017: Rework the Fibonacci example (#107132)
python/cpython: Doc/tutorial/introduction.rst

First Steps Towards Programming
===============================

We can use Python for more complex tasks than adding two and two together.
For instance, we can compute the numbers of the
`Fibonacci series <https://en.wikipedia.org/wiki/Fibonacci_number>`_.
To do that we need to utilize three following concepts.

Multiple Assignment
-------------------

A *multiple assignment* of variables us allows to set values of more than one
variable on a single line.  Notice that if the value assignment is a

> expression, it is first evaluated and then assigned.::

Reviewer: `.::` will produce a full-stop _and_ a colon:

> .:

One or the other. I suggest losing the full-stop.

   >>> a, b = 1, 5  # multiple assignment of two variables
   >>> a
   0
   >>> b
   5
   >>> a, b = b, a + b  # first a is set to value of b, then b is set to sum of a and b
   >>> a
   5
   >>> b
   6

Repeating our code
------------------

The :keyword:`while` loop is example of a repeating cycle.  It performs

> a block of code over as long as the condition (in example: ``a < 10``) is

Reviewer: "in example" > "for example"

Author: What I wanted to say is that "in the example below the condition is the following expression `a < 10`", but that is no longer present in the examples. I will rework that.

fulfilled.  The test used in the example is a simple comparison you might
know from arithmetics.  Fulfilled condition have the value True.  Unfulfilled
condition have the value False.::

   >>> 1 < 3  # 1 is less than 3
   True
   >>> 'king' == "king"  # the strings are equal
   True
   >>> 4 >= 6  # 4 is not greater than or equal to 6
   False

In Python, the following comparison operators are used:

  * ``<``: Less than
  * ``>``: Greater than
  * ``==``: Equal to
  * ``<=``: Less than or equal to
  * ``>=``: Greater than or equal to
  * ``!=``: Not equal to

::

   >>> count = 0;  # define variable to which we will be adding 1 in a loop
   >>> while count < 5: count = count + 1  # hit enter one more time to start the loop
   ...
   >>> count  # when count reached value of 5, while loop finished
   5

> Note that block inside the while loop, or *body* of the loop is *indented*.

Reviewer: I suggest using ``while``, to draw attention to it.

Indentation is Python's way of grouping statements together.  When you use
the Python shell, you need to type a tab or space(s) for each indented line.
Each line within a block must be indented by the same amount.
When we want to exit the indented block of code in the Python shell, we must

> follow it by a blank line to indicate its completion.  This way, parser knows

Reviewer: "... the parser"

> we have finished typing the last line).::

Reviewer: Again, here we have a full-stop _and_ a colon in the output.

   >>> number = 1
   >>> while number < 5:
   ...     print(number)
   ...     number = number + 1  # increase the value by one with each repetition
   ...
   1
   2
   3
   4

Function argumets
-----------------

We already know :func:`print` function, that writes the value of the
argument(s) it receives on screen.  The arguments are enclosed within
parentheses ``()``. In simplest form, like ``print(a, b)`` the arguments
are positional,  meaning the function processes them in the same order
as they are written.::

   >>> group = "Knights"
   >>> say_what = '"Ni!"'
   >>> print(group, 'Who Say', say_what)
   Knights Who Say "Ni!"

The ``print()`` function also offers a useful keyword argument called *end*,
which can be used to avoid the newline after the output, or end the output
with a different string::

   >>> count = 0
   >>> count = 10
   >>> while count > 5:
   ...     print(count, end=' ')  # end the output of print() with a whitespace
   ...     count = count - 1  # lower the value by one with each repetition
   ...

>    10 9 8 7 6

Reviewer: This is an elegant touch.

Author: Thank you :)

Fibonacci series
----------------

Great!  Now that we have the knowledge, let's write code to print all numbers
from the Fibonacci series that are lower than 150.

   >>> a, b = 0, 1
   >>> while a < 150:
   ...      print('', a, '', end='->')
   ...      a, b = b, a+b
   ...

>     0 -> 1 -> 1 -> 2 -> 3 -> 5 -> 8 -> 13 -> 21 -> 34 -> 55 -> 89 -> 144 ->

Reviewer: And I like the way that the example illustrates the _value_ of using `end` - no need to explain, the example does it much more effectively. This will stick in someone's mind.

First Steps Towards Programming
===============================

We can use Python for more complex tasks than adding two and two together.
For instance, we can compute the numbers of the
`Fibonacci series <https://en.wikipedia.org/wiki/Fibonacci_number>`_.

> To do that we need to utilize three following concepts.

Reviewer 2: ```suggestion
To do that we need to use the three following concepts.
```

Let's use the [plain English](https://www.plainlanguage.gov/guidelines/words/use-simple-words-phrases/) "use" instead of "utilize".

Multiple Assignment
-------------------

> A *multiple assignment* of variables us allows to set values of more than one

Reviewer 2: ```suggestion
A *multiple assignment* of variables us allows to set the values of more than one
```

> variable on a single line.  Notice that if the value assignment is a

Reviewer 2: ```suggestion
variable on a single line.  Notice that if the value assignment is an
```

expression, it is first evaluated and then assigned::

   >>> a, b = 1, 5  # multiple assignment of two variables
   >>> a

>    0

Reviewer 3: Needs to be `1`
![image](https://github.com/python/cpython/assets/23421705/900d91f6-44cd-46e8-9e27-b4567d91e9d8)

Author: Is this a doctest bug?

   >>> b
   5
   >>> a, b = b, a + b  # first a is set to value of b, then b is set to sum of a and b
   >>> a
   5
   >>> b
   6

Repeating our code
------------------

> The :keyword:`while` loop is example of a repeating cycle. It performs

Reviewer 2: ```suggestion
The :keyword:`while` loop is an example of a repeating cycle. It performs
```

> The :keyword:`while` loop is example of a repeating cycle.  It performs

Reviewer 3: ```suggestion
The :keyword:`while` loop is example of a repeating cycle. It performs
```

> a block of code over as long as the condition is

Reviewer 3: ```suggestion
a block of code continuously as long as the condition is
```

Reviewer: I think that "over and over again" would be better than both these. "Continuously" is technically correct, but implies continuity, as in a continuous stream of water - whereas what we need to refer to is repeated execution of distinct, discrete blocks of code.

Author: Yes, this is more precise and clearer!

fulfilled.  The test used in the example is a simple comparison you might

> know from arithmetics.  Fulfilled condition have the value ``True``.

Reviewer 3: ```suggestion
know from arithmetics.  Fulfilled conditions have the value ``True``.
```

> Unfulfilled condition have the value ``False``::

Reviewer 3: ```suggestion
Unfulfilled conditions have the value ``False``::
```

   >>> 1 < 3  # 1 is less than 3
   True
   >>> 'king' == "king"  # the strings are equal
   True
   >>> 4 >= 6  # 4 is not greater than or equal to 6
   False

In Python, the following comparison operators are used:

  * ``<``: Less than
  * ``>``: Greater than
  * ``==``: Equal to
  * ``<=``: Less than or equal to
  * ``>=``: Greater than or equal to
  * ``!=``: Not equal to

Let's create a simple loop.  We need the keyword ``while`` and a condition;
in this case, ``count < 5``.  To ensure that the loop finishes, we must make
sure that the condition is *not fulfilled* at a certain step.  Otherwise, the
code would repeat indefinitely.  To achieve that, we can increase the value

> of count.  As soon as the variable ``count`` reaches the value 5, the

Reviewer 2: ```suggestion
of ``count``.  As soon as the variable ``count`` reaches the value 5, the
```

condition will be ``False``::

>    >>> count = 0;  # define variable to which we will be adding 1 in a loop

Reviewer 2: ```suggestion
   >>> count = 0  # define variable to which we will be adding 1 in a loop
```

>    >>> while count < 5: count = count + 1  # hit enter one more time to start the loop
>    ...
>    >>> count  # when count reached value of 5, while loop finished
>    5
> 
> Note that block inside the ``while`` loop, or *body* of the loop is *indented*.

Reviewer 2: The example above uses a one-liner: `while count < 5: count = count + 1`

Let's use indentation in the example, to match the text. Something like this?

```suggestion
   >>> while count < 5:
   ...    count = count + 1  # hit enter one more time to start the loop
   ...
   >>> count  # when count reached value of 5, while loop finished
   5

Note that block inside the ``while`` loop, or *body* of the loop is *indented*.
```

Indentation is Python's way of grouping statements together.  When you use
the Python shell, you need to type a tab or space(s) for each indented line.
Each line within a block must be indented by the same amount.
When we want to exit the indented block of code in the Python shell, we must
follow it by a blank line to indicate its completion.  This way, the parser
knows we have finished typing the last line)::

   >>> number = 1
   >>> while number < 5:
   ...     print(number)
   ...     number = number + 1  # increase the value by one with each repetition
   ...
   1
   2
   3
   4

> Function argumets

Reviewer 4: Typo here

> Function argumets

Author: ```suggestion
Function arguments
```

> -----------------

Author: ```suggestion
------------------
```

> We already know :func:`print` function, that writes the value of the

Reviewer 4: We already know *the* print function ?

> We already know :func:`print` function, that writes the value of the

Author: ```suggestion
We already know the : func:`print` function, which writes the value of the
```

> We already know the : func:`print` function, which writes the value of the

Author: ```suggestion
We already know the :func:`print` function, which writes the value of the
```

argument(s) it receives on screen.  The arguments are enclosed within

> parentheses ``()``. In simplest form, like ``print(a, b)``, the arguments
> are positional,  meaning the function processes them in the same order

Reviewer 2: ```suggestion
parentheses ``()``. In its simplest form, like ``print(a, b)``, the arguments
are positional, meaning the function processes them in the same order
```

as they are written::

   >>> group = "Knights"
   >>> say_what = '"Ni!"'
   >>> print(group, 'Who Say', say_what)
   Knights Who Say "Ni!"

The ``print()`` function also offers a useful keyword argument called *end*,
which can be used to avoid the newline after the output, or end the output
with a different string::

   >>> count = 0
   >>> count = 10
   >>> while count > 5:
   ...     print(count, end=' ')  # end the output of print() with a whitespace

>    ...     count = count - 1  # lower the value by one with each repetition

Reviewer 3: ```suggestion
   ...     count = count - 1  # decrement the value each iteration
```

Reviewer: I would strongly prefer "decrease the value by one".

The reason is that many people will not be familiar with the word "decrement", it's used very little in English outside technical contexts.

   ...
   10 9 8 7 6

Fibonacci series
----------------

Great!  Now that we have the knowledge, let's write code to print all numbers
from the Fibonacci series that are lower than 150::

   >>> a, b = 0, 1
   >>> while a < 150:

>    ...      print('', a, '', end='->')

Reviewer 3: I understand that this utilities the print sep param, however this is nor explained nor inferred, so to a beginner it may be confusing 

Perhaps it could be beneficial to show off f-strings like 
```suggestion
   ...      print(f'{a} ->', end='')
```
Or to keep end as `->`
```suggestion
   ...      print(f'{a} ', end="->")
```

Reviewer: I very strongly prefer the original `print('', a, '', end='->')`, which reduces the punctuation the learner has to deal with to a minimum, and emphasises instead the connection between more easily assimilated things, like the English word "end" and the "arrow" ("->"). 

Showing off f-strings while demonstrating a different functionality introduces unnecessary cognitive burden. Two different kinds of quotes and two different kinds of brackets requires a lot of parsing and stopping. I don't think this is the right solution here.

Author: My reasoning: As this is an Introduction chapter, I wanted to keep the number of introduced concepts to the minimum. There are ways to code this better, but it is not formally wrong. The default separator is already shown in the example above and this just builds on that, so it's hopefully not confusing to the reader.

>    ...      a, b = b, a+b

Reviewer 3: ```suggestion
   ...      a, b = b, a + b
```
Looks better and I believe its what black would format

guess when you have typed the last line).  Note that each line within a basic
block must be indented by the same amount.::

   >>> number = 1
   >>> while number < 5:
   ...     print(number)
   ...     number = number + 1
   ...
   1
   2
   3
   4

Function argumets
-----------------

We already know :func:`print` function, that writes the value of the
argument(s) it is given on screen.  The arguments are specified inside
parentheses ``()``. In simplest form, f.e. ``print(a, b)`` the arguments
are positional.  That means function process them in same order we wrote
them.::

   >>> group = "Knights"
   >>> say_what = '"Ni!"'
   >>> print(group, 'Who Say', say_what)
   Knights Who Say "Ni!"

The keyword argument *end* can be used to avoid the newline after the output,
or end the output with a different string::

   >>> count = 0
   >>> count = 10
   >>> while count > 5:
   ...     print(count, end=' ')
   ...     count = count - 1
   ...
   10 9 8 7 6

Fibonacci series
----------------

We now know all we need to write code that will print all numbers from the
Fibonacci series! Let's check what nubers in it are lower than 1000.

     >>> a, b = 0, 1
     >>> while a < 1000:

>      ...     print(a, end=', ')

Reviewer 2: Do we want the space in `end` in this example? If so, update the output on line 546 with spaces after commas, so doctest passes on the CI.

Author: Thank you for your feedback.  I did this to showcase a different output style compared to the previous example, which used one character. My hope it might support a desire to experiment further with the end keyword. Maybe we might come up with a better way to do it. Or just not mess with it here 😂.

Author: In 1fae12b I tried a bit more obvious "two characters" and concepts from another example to create maybe a more interesting final example. Let me know what you think.
